added support for stripping output from Zeppelin Notebooks

ankitrokdeonsns opened the pull request:
Zeppelin Notebooks are popular especially while working with Apache Spark Framework.
Since the basic structure of the IPython / Jupyter and Zeppelin Notebooks is similar it would be cool to provide nbstripout support for Zeppelin Notebooks as well.

kynan commented:
Thank you for the pull request. There's a few points to discuss:

1.  Your PR only addresses the code path of calling `nbstripout` with a file argument. There is also the (more common) way of reading from stdin, which is used in the `git` filter.

1.  To handle the case of reading from stdin we need to either
    * detect the file type or
    * specify the file type using a flag

    My preference is being explicit using a flag e.g. `--format` which defaults to `jupyter`

1. We'd need to decide how to handle the various flags `nbstripout` supports. Do they have equivalents for Zeppelin? Does Zeppelin have execution counts? Or notebook and cell level metadata? Tags?

    If it doesn't make sense to support these flags, what do we want to do when running in "Zeppelin mode" and an incompatible flag is passed? Ignore silently? Ignore with a warning? Fail?

Please also add tests and update the documentation as part of this pr.

ankitrokdeonsns commented:
> Thank you for the pull request. There's a few points to discuss:
> 
> 1. Your PR only addresses the code path of calling `nbstripout` with a file argument. There is also the (more common) way of reading from stdin, which is used in the `git` filter.

Correct. Zeppelin notebooks (v0.9+) have `*.zpln` extension which can be used for setting up the git filter.
Will be doing so in upcoming commits.

> 2. To handle the case of reading from stdin we need to either
>    
>    * detect the file type or
>    * specify the file type using a flag
>    
>    My preference is being explicit using a flag e.g. `--format` which defaults to `jupyter`

`.zpln` extension should take care of this. For older Zeppelin notebooks it would be a little more tricky since they are stored as `note.json` files

> 3. We'd need to decide how to handle the various flags `nbstripout` supports. Do they have equivalents for Zeppelin? Does Zeppelin have execution counts? Or notebook and cell level metadata? Tags?
>    If it doesn't make sense to support these flags, what do we want to do when running in "Zeppelin mode" and an incompatible flag is passed? Ignore silently? Ignore with a warning? Fail?
> 

Yes some similar metadata is present like last execution time. I will update more on this.

> Please also add tests and update the documentation as part of this pr.

Will do.

ankitrokdeonsns commented:
As far as I know Zeppelin does not have support for adding metadata or tags to cells. So `keep-output` is currently not possible for Zeppelin notebooks.
Zeppelin notebooks also do not have execution count. So `keep-count` also cant be supported.
We can ignore such things silently for zeppelin notebooks. At least for now.

ankitrokdeonsns commented:
Structure of Zeppelin Notebook cell:
```
{
      "text": "%md\n# Hello",
      "user": "anonymous",
      "dateUpdated": "2020-07-08 16:46:54.732",
      "config": {
        "colWidth": 12.0,
        "fontSize": 9.0,
        "enabled": true,
        "results": {},
        "editorSetting": {
          "language": "text",
          "editOnDblClick": false,
          "completionKey": "TAB",
          "completionSupport": true
        },
        "editorMode": "ace/mode/text"
      },
      "settings": {
        "params": {},
        "forms": {}
      },
      "results": {},
      "apps": [],
      "progressUpdateIntervalMs": 500,
      "jobName": "paragraph_1594203861277_1152517137",
      "id": "paragraph_1594203861277_1152517137",
      "dateCreated": "2020-07-08 15:54:21.277",
      "dateStarted": "2020-07-08 16:46:54.740",
      "dateFinished": "2020-07-08 16:46:54.854",
      "status": "FINISHED"
    }
```
We could consider to clear keys such as `dateCreated`, `dateStarted`, `dateFinished`, `status` while we clear the output. What do you think?

kynan commented:
That sounds reasonable! I'm still in favor of having a flag to `--format` which defaults to `jupyter` and also supports `zeppelin`, since that will work both with file input and reading from stdin.

ankitrokdeonsns commented:
Ok. Currently In My solution `--format` is not necessarily required but we could use that for supporting older versions of zeppelin notebooks. I will add that flag as you suggested. 👍

kynan commented on nbstripout/_nbstripout.py:
```diff
@@ -199,7 +204,7 @@ def uninstall(git_config, user=False, attrfile=None):
     if path.exists(attrfile):
         with open(attrfile, 'r+') as f:
-            lines = [line for line in f if not (line.startswith('*.ipynb filter') or line.startswith('*.ipynb diff'))]
+            lines = [line for line in f if not (line.startswith('*.ipynb filter') or line.startswith('*.zpln filter') or line.startswith('*.ipynb diff'))]
```
```suggestion
            patterns = ('*.ipynb filter', '*.zpln filter', '*.ipynb diff')
            lines = [line for line in f if not any(line.startswith(p) for p in patterns))]
```

kynan commented on nbstripout/_utils.py:
```diff
@@ -1,6 +1,6 @@
 import sys
 
-__all__ = ["pop_recursive", "strip_output", "MetadataError"]
+__all__ = ["pop_recursive", "strip_output", "strip_zeppelin_output","MetadataError"]
```
```suggestion
__all__ = ["pop_recursive", "strip_output", "strip_zeppelin_output", "MetadataError"]
```

ankitrokdeonsns commented:
Sure. I was busy last week. I will add the test tomorrow and update the pull request.

ankitrokdeonsns commented:
I have incorporated changes as you suggested. Added tests and updated documentation. I believe now it is ready for merge.
Do let me know if I missed anything.

ankitrokdeonsns commented:
Hi @kynan I believe this PR is ready for merge.

kynan commented on nbstripout/_nbstripout.py:
```diff
@@ -287,6 +298,10 @@ def main():
                       help='Print version')
     parser.add_argument('--force', '-f', action='store_true',
                         help='Strip output also from files with non ipynb extension')
+    parser.add_argument('--mode'
```
Set the default to `jupyter`. Also please format according to PEP 8

ankitrokdeonsns replied:
done

kynan commented on nbstripout/_nbstripout.py:
```diff
@@ -307,6 +322,12 @@ def main():
+    if args.mode is None:
+        args.mode = 'jupyter'
+    elif args.mode not in ['jupyter', 'zeppelin']:
+        print('mode must be from [jupyter, zeppelin]')
```
Raise an exception instead. A library should never `exit`.

ankitrokdeonsns replied:
done

kynan replied:
Actually, that's not needed at all if you use `choices` for the argument.

kynan reviewed (changes requested):
Please rebase.

kynan commented on nbstripout/_nbstripout.py:
```diff
@@ -253,6 +261,17 @@ def install(git_config, install_location=INSTALL_LOCATION_LOCAL, attrfile=None):
+            f.write('\n')
+        if not filt_exists:
+            print('*.ipynb filter=nbstripout', file=f)
+        if not zeppelin_filt_exists:
```
This rebase seems to have gone wrong: please move the Zeppelin related option to line 254

kynan commented on nbstripout/_nbstripout.py:
```diff
@@ -420,6 +442,11 @@ def main():
         'cell.metadata.hidden',
         'cell.metadata.scrolled',
     ]
+    if args.mode is None:
```
This is no longer needed

ankitrokdeonsns replied:
Good catch. I dont know how I missed this. Thanks a lot.

kynan commented on nbstripout/_nbstripout.py:
```diff
@@ -379,6 +399,8 @@ def main():
                         help='Strip output also from files with non ipynb extension')
     parser.add_argument('--max-size', metavar='SIZE',
                         help='Keep outputs smaller than SIZE', default='0')
+    parser.add_argument('--mode', '-m', default='jupyter',
```
use `choices=['jupyter', 'zeppelin']`

ankitrokdeonsns replied:
done

kynan reviewed (changes requested):
Please rebase again. I've made some changes to the GitHub actions config while you were rebasing, sorry!

ankitrokdeonsns commented:
incorporated latest requested changes

kynan reviewed (changes requested):
Nearly there I think!

ankitrokdeonsns commented:
Good catch @kynan. I think this remained due to bad merging during rebase. Thanks. Deleted the extra block.

kynan reviewed (approved):
Thanks!
